File: four_languages/poker/java/Poker/src/poker/Poker.java
package poker;

import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.TreeMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;


/* Poker in java, 
  input one hand parse the hand for score and a List 
	poker.scoreHand() => PokerResult object
	PokerResult = { int score, List valuesRanks  }

	BestHand will compare two PokerResult objects 
	=> return a1 < a2 => -1, a1==a2 => 0; a1>a2 => 1
*/
/*
  A big logic mistake in scoreHand(), it has to be if else, if else
  not just if statement, then the score (straight && flushes) = 8 
  then be modified in later if statement to score = 4 (straight)
*/
public class Poker {
	public static final String RANKS_INDEX = "--23456789TJQKA";
	public static final Integer[] SPECIAL_STRAIGHT = new Integer[]{ 2,3,4,5,14};
	
	String hand;
	private char suits[];	

	// it creates in createRanksHash()
	private Integer[] ranksSorted = new Integer[5];	

	// store rank as key, value is count of rank
	private HashMap<Integer, Integer> ranks_hash = new HashMap<>();

	// use to create PokerResult object, this should be reset to 0 in scoreHand()
	int score = 0;	

	Poker(String handStr) {
		// BestHand class will pass in each hand	
		hand = handStr.replace("10", "T");	// replace 10 to 'T'
		suits = parseHand(hand, true);		// get suits
		createRanksHash(hand);	// create ranks_hash
		// scoreHand(); //for testing Poker.java by itself
	}

	// parse hand to create  suits and ranks char[], 
	// but only returns one of them
	// if argument flag = true => suits char[], flag = false => ranks char[]
	private char[] parseHand(String handStr, boolean flag) {
		char[] suits = new char[5];
		char[] ranks = new char[5];

		// take out spaces
		String str = handStr.replaceAll(" ", "");
		
		// create two char[], one for suits and one for ranks
		char[] handAry = str.toCharArray();
		int idx1 = 0;
		int idx2 = 0;
		for( int i=0; i < handAry.length ; i++ ) {
			if ( (i % 2) != 0) {
				suits[idx1++] = handAry[i];
			} else {
				ranks[idx2++] = handAry[i];
			}
		}
		// depend on flag return suits char[] or ranks char[]
		if (flag) {
			return suits;
		} else {
			return ranks;		
		}
	}

	// create ranks_hash, key is rank, and value is count of the rank
	private void createRanksHash(String handStr) {

		char[] ranks = parseHand(handStr, false);		// [ 6,5,K,5,T]
		
		// convert ranks 'T', 'J', 'Q', 'A' into values 
		for( int i=0; i < ranks.length; i++) {
			ranksSorted[i] = RANKS_INDEX.indexOf(ranks[i]);
		}	
		Arrays.sort(ranksSorted);	//[ 5,5,6,10,13]

		// create ranks hash with key= rank , value = count of rank
		for(int i=0; i < ranksSorted.length; i++) {
			if (ranks_hash.containsKey(ranksSorted[i])) {
				ranks_hash.put(ranksSorted[i], ranks_hash.get(ranksSorted[i]) + 1);
				//System.out.println(ranks_hash.get(ranksSorted[i]));	// => 2
			} else {
				ranks_hash.put(ranksSorted[i], 1);
			}
		}
	}

	/*
	poker.scoreHand() uses hand, suits, ranks_hash
	it calculate score by
	- check for flush, special straight, straight, kinds
	- get int score 0 - 8

	create HashMap ranks_hash
	- with key= rank and value= rank's count

	creates List valuesRanks 
	- ranks_hash reverse sort by ranks
	- then reverse sort by values
	=> List valuesRanks

	create PokerResult object
	return PokerResult = { int score, List valuesRanks }
	*/
	public PokerResult scoreHand() {
		score = 0;	// should be reset for each hand

		// check for flush, work => false
		boolean flush = true;
		for(int i=1; i< suits.length; i++) {
			if (suits[i] != suits[i-1]) {
				flush = false;
				break;
			}
		}

		// check for special straight [ 2,3,4,5,14];
		// so we can modify ranks_hash
		if ( Arrays.equals(ranksSorted, SPECIAL_STRAIGHT)) {
			// then change 14 to 1, change ranks_hash => {1: 1, 2:2, 3:3, 4:4, 5:5}
			ranks_hash.clear();		//k I forgot to clear the hash 
			ranks_hash.put(1,1);
			ranks_hash.put(2,2);
			ranks_hash.put(3,3);
			ranks_hash.put(4,4);
			ranks_hash.put(5,5);

			// need to change ranksSorted from
			//[ 2,3,4,5,14]; to [1,2,3,4,5]
			ranksSorted = new Integer[]{1, 2, 3, 4, 5}; 
		}

		// check for straight
		boolean straight = true;
		for(int i=1; i < ranksSorted.length; i++) {
			if (ranksSorted[i] != ranksSorted[i-1] + 1) {
				straight = false;
				break;
			}
		}
		
		// create kinds = values of ranks_hash, sort in reverse
		Collection<Integer> kinds_col = new ArrayList<>();
		kinds_col = ranks_hash.values();
		//System.out.println(Arrays.asList(ranks_hash));
		Integer[] kinds = new Integer[kinds_col.size()];
		Object[] kinds2 = kinds_col.toArray();
		
		// cast Object into int
		for(int i=0; i<kinds2.length; i++) {
			kinds[i] = (int) kinds2[i];
			//System.out.println(kinds[i]);
		}
		Arrays.sort(kinds, Collections.reverseOrder());	// reverse array order, [2,1,1,1]
		//System.out.println(Arrays.toString(kinds));

		Integer[] score_7 = new Integer[]{4,1};
		Integer[] score_6 = new Integer[]{3,2};
		Integer[] score_3 = new Integer[]{3,1,1};
		Integer[] score_2 = new Integer[]{2,2,1,};
		Integer[] score_1 = new Integer[]{2,1,1,1};

		// calculate score: kinds, flush, straight
		if (flush && straight) { 
			score = 8;
		} else if (Arrays.equals(kinds, score_7)) { 
			score = 7;
		} else if (Arrays.equals(kinds, score_6)) { 
			score = 6;
		} else if (flush) { 
			score = 5;
		} else if (straight) { 
			score = 4;
		} else if (Arrays.equals(kinds, score_3)) { 
			score = 3;
		} else if (Arrays.equals(kinds, score_2)) { 
			score = 2;
		} else if (Arrays.equals(kinds, score_1)) { 
			score = 1;
		} else {
			score = 0;
		}

		//k need to reverse ranks sort on ranks_hash 
		// then sort by values 

		// reverse sort on ranks => ordered List result1 on ranks
		List result1 = entriesSortedByValues(ranks_hash, true);
		//System.out.println(result1);	// [11=1, 8=2, 2=2]

		// sort the ordered List by value => List valuesRanks
		List valuesRanks = new ArrayList<Map.Entry<Integer, Integer>>();
	
		Map.Entry<Integer, Integer> entry1;
		Map.Entry<Integer, Integer> cur_entry;
		Integer  cur_value, value1;

		entry1 = (Map.Entry<Integer, Integer>) result1.get(0);
		value1 = entry1.getValue();
		valuesRanks.add(0, entry1);

		for (int i=1; i< result1.size(); i++) {
			cur_entry = (Map.Entry<Integer, Integer>) result1.get(i);
			cur_value = cur_entry.getValue();
			if ( cur_value >= value1 ) {
				// insert this Map.Entry before entry1, set valu1 = cur_value
				valuesRanks.add(i-1, cur_entry);
				value1 = cur_value;
			} else  {
				valuesRanks.add(i, cur_entry);
			}
		}
		System.out.println(valuesRanks);	// [8=2, 2=2, 11=1], [5=2, 4=2, 8=1]
		System.out.println("score: " + score);
		PokerResult result = new PokerResult(score, valuesRanks);

		return result;	
	}
	
	public static void main(String[] args) {
		String hand = "2S 8H 2S 8D JH";
		new Poker(hand);
	} 
	
	// compare two map with duplicated values == values are not unique 
	// flag = true sort reverse key, flag = false sort reverse value 
	static <K, V extends Comparable<? super K>> 
			List<Map.Entry<Integer,Integer>> entriesSortedByValues(Map<Integer,Integer> map, boolean flag) {
		boolean flag_K_V = flag;
		List<Map.Entry<Integer,Integer>> sortedEntries = new ArrayList<Map.Entry<Integer,Integer>>(map.entrySet());

		Collections.sort(sortedEntries, 
			new Comparator<Map.Entry<Integer,Integer>>() {
				@Override
				public int compare(Map.Entry<Integer,Integer> e1, Map.Entry<Integer,Integer> e2) {
					if (flag_K_V ) {
						return e2.getKey().compareTo(e1.getKey());
					} else {
						//k I didn't use this part
						return e2.getValue().compareTo(e1.getValue());
					}
				}
			}
		);
		return sortedEntries;
	}
}

File: four_languages/poker/javascript/poker_jest/poker.js
//poker.js 11/27/18

const RANKS_INDEX = '--23456789TJQKA' 

// return 0:equal; 1: a> b; -1: a < b
// assume a and b have same length
// this compArray() uses to compare ranks when the two hands have the same score.
const compArray = (a, b)  => {
	for (let i=0; i< a.length; i++) {
		let flag = a[i] - b[i]
		if (flag > 0) 
			return 1
		else if (flag < 0) {
			return -1
		}
	}
	return 0
}

// compKind(a, b) return a === b => true; otherwise => false
const compKind = (a, b) => {
	if (a.length !== b.length) {
		return false
	}
	for (let i=0; i < a.length; i++) {
		if (a[i] !== b[i]) {
			return false
		}
	}
	return true
}

// return 0:equal; 1: a> b; -1: a < b
// compare Map objects, compare the keys of the Map objects not the values
const compMap = (map1, map2) => {
	// get keys of two maps, then compare arrays is easier
	const keys1_iterator = map1.keys()
	const keys2_iterator = map2.keys()
	const keys1 = Array.from(keys1_iterator)
	const keys2 = Array.from(keys2_iterator)
	
	if ( compArray(keys1, keys2) > 0) {
		return 1
	} else if (compArray(keys1, keys2) < 0) {
		return -1
	} else {
		return 0
	}		
}

// find the besthand or tie from an array of hands
const bestHands = hands => {
	let best_hands = hands[0]
	const pk1 = new Poker(hands[0])
	let best_score_map = pk1.score_hand() // => [score, Map object]
	let best_score = best_score_map[0]
	let best_map = best_score_map[1]
	
	// parse each hand
	for(let i=1; i<hands.length; i++) {
		const hand = hands[i]
		let pk = new Poker(hand)
		
		// get score and map object from poker.score_hand()
		const score_map = pk.score_hand()
		const score = score_map[0]
		const map = score_map[1]
		
		// compare score
		if (best_score < score) {
			best_hands = hand
			best_score = score
		} else if (best_score === score) {
			// need to compare ranks map
			const flag = compMap(best_map, map) 	// only place use compMap()
			if (flag < 0) {
				best_hands = hand
				best_score = score
			} else if (flag === 0) {
				// it is tie, add hand to best_hands
				// change best_hands from String to String Array
				best_hands = [best_hands].concat(hands[i])
			}
		} 
	}
	return best_hands
}

class Poker {   
	constructor(hand) {
		this.hand = hand
		this.suits = this.parseHand(true)		// get suits
		this.ranks = this.parseHand(false)	// get ranks
		
		// ranks_map: keys= ranks, values = ranks count
		// ranks_map is reverse sorted by ranks(key), ex: [ {11:1}, {8:2}, {2:2}]
		this.ranks_map = this.create_ranks_map()
	}
	
	// => ['2', 'S', '4', 'H', '6', 'S', '4', 'D', 'J', 'H']
	get_hand_ary() {
		const hand = this.hand
		const hand1 = hand.replace(/\s/g,'')
		const hand2 = hand1.replace(/10/g,'T')     // '10' to 'T'
		const hand_ary = hand2.split('')
		return hand_ary
	}

	parseHand(flag) {
		const hand_ary = this.get_hand_ary()
		let suits = []	// => ["S", "H", "S", "D", "H"]
		let ranks = []	// => ["5", "2", "4", "A", "3"]
		hand_ary.map( (elem, idx) => {
			 if ( idx%2 === 0) { // works too
				ranks.push(hand_ary[idx])
			} else {
				suits.push(hand_ary[idx])
			}
		})
		if (flag) {
			return suits
		} else {
			return ranks
		}
	}
	
	// ranks = ["2", "8", "2", "8", "2"]
	// => ranks_map = [ {11:1}, {8:2}, {2:2} ] a map with keys reverse sorted
	create_ranks_map() {
		let ranks_ary = []
		// convert ranks chars to its index values
		this.ranks.map( e => {
			ranks_ary.push(RANKS_INDEX.indexOf(e))
		})
		
		// => [11,8,8,2,2]  is reverse sort of ranks array
		ranks_ary.sort( (a,b) => b - a ) 
		
		// create ranks_map with reverse sorted ranks key, value = rank count
		let ranks_map = new Map()
		ranks_ary.forEach( key => {
			// if key exists
			if (ranks_map.has(key)) {
				ranks_map.set(key, ranks_map.get(key) + 1)
			} else {
				ranks_map.set(key, 1)
			}
		})
		
		// key in ranks_map is already key reverse sorted, ex: [{11:1}, {8:2}, {2:2}]
		// when calculate the ranks score, I will need to reverse sort by volue
		return ranks_map
	}
	
	score_hand() {
		// check for flush
		const flush = this.suits.every( e => this.suits[0] === e )
		
		// create sorted_ranks in value, sorted by ASC
		// get keys from ranks_map, then reverse sort keys from DESC to ASC
		const sortedIterator = this.ranks_map.keys()  // => iterator 
		// convert iterator to Array
		let sorted_ranks = Array.from(sortedIterator).reverse()	// change from DESC to ASC for straight comparison
		
		// check for special straight [2,3,4,5,14] => modify this.rank_map
		if (compKind(sorted_ranks, [2,3,4,5,14]) ) {
			// console.log("special straight")
			this.ranks_map = [ [1,1], [2,1], [3,3], [4,4], [5,5] ]
			sorted_ranks = [1,2,3,4,5]
		}
		
		// check straight
		let straight = true
		for ( let i=0; i<sorted_ranks.length; i++) {
			if ( sorted_ranks[i] !== sorted_ranks[0] + i) {
				straight = false
				break
			}       
		}
		
		// ranks_map is already key reverse sorted, ex:[ {11,1}, {8:2}, {2:2} ]
		// create ranks_values Map = reverse sort values on ranks_map
		// ranks_values Map will be included in return statement
		
		//const ranks_values = Array.from(this.ranks_map.entries()).sort((a,b) => b[1] - a[1])  // works
		const ranks_values = new Map([...this.ranks_map.entries()].sort((a,b) => b[1] - a[1]) )
		
		// create kinds array
		const values_map = ranks_values.values()			// => Iterator of values
		const kinds = Array.from(values_map)					// convert Iterator to array of values, [4,1], [2,1,1,1]
		
		// calculate score: kinds, flush, straight
		
		let score = 0
		if (flush && straight) score = 8
		else if ( compKind(kinds, [4,1]) ) score = 7
		else if ( compKind(kinds, [3,2]) ) score = 6
		else if ( flush ) score = 5
		else if ( straight ) score = 4
		else if ( compKind(kinds, [3,1,1]) ) score = 3
		else if ( compKind(kinds, [2,2,1]) ) score = 2
		else if ( compKind(kinds, [2,1,1,1]) ) score = 1
		else score = 0
		
		return [score].concat(ranks_values)		// ranks_values is a Map object
	}
}

export { bestHands }
